### PR TITLE
Scoped packages 403 on yarn install with empty cache

### PR DESCRIPTION
**Pull request by lukeggchapman**
**Summary**

Fix scoped packages not adding correct auth headers (#4016)

**Test Plan**

Having set up config for a scoped registry with an auth token. It should be possible to run `yarn cache clean` followed by `yarn install` without having the scoped packages 403

### CODE REVIEW

File: src/registries/npm-registry.js [hunk 1]
```diff
@@ -48,6 +48,11 @@ function isPathConfigOption(key: string): boolean {
   return PATH_CONFIG_OPTIONS.indexOf(key) >= 0;
 }
 
+function isScopedPackage(packageIdent: string): boolean {
+  // scoped package names will begin with '@', scoped path names will contain '/@'
+  return !!packageIdent.match(/^@|\/@/);
```
Comment (BYK) [inline src/registries/npm-registry.js]: Use `/^@|\/@/.test(packageIdent)` instead please.

File: src/registries/npm-registry.js [hunk 2]
```diff
@@ -218,21 +221,27 @@ export default class NpmRegistry extends Registry {
     }
   }
 
-  getScope(packageName: string): string {
-    return !packageName || packageName[0] !== '@' ? '' : packageName.split(/\/|%2f/)[0];
+  getScope(packageIdent: string): string {
+    // Matches '@scope' in a packageName or pathname, otherwise returns ''
+    if (packageIdent && isScopedPackage(packageIdent)) {
```
Comment (BYK) [inline src/registries/npm-registry.js]: No need for the `isScopedPackage` check since the below matcher will fail anyways.

File: src/registries/npm-registry.js [hunk 3]
```diff
@@ -218,21 +221,27 @@ export default class NpmRegistry extends Registry {
     }
   }
 
-  getScope(packageName: string): string {
-    return !packageName || packageName[0] !== '@' ? '' : packageName.split(/\/|%2f/)[0];
+  getScope(packageIdent: string): string {
```
Comment (BYK) [inline src/registries/npm-registry.js]: This function better written as:
```
getScope(packageIdent: string): string {
    const match = packageIdent.match(/(^@|(?!\/)@).+?((?=%2f)|(?=\/))/);
    return match && match[1] || '';
}

File: src/registries/npm-registry.js [hunk 4]
```diff
@@ -218,21 +221,27 @@ export default class NpmRegistry extends Registry {
...
+    // Matches '@scope' in a packageName or pathname, otherwise returns ''
+    if (packageIdent && isScopedPackage(packageIdent)) {
+      const matched = packageIdent.match(/(^@|(?!\/)@).+?((?=%2f)|(?=\/))/);
+      return (matched && matched[0]) || '';
```
Comment (BYK) [inline src/registries/npm-registry.js]: You want `matched[1]` if you want the first group, not `matched[0]`

File: src/registries/npm-registry.js [hunk 5]
```diff
@@ -218,21 +221,27 @@ export default class NpmRegistry extends Registry {
     }
   }
 
-  getScope(packageName: string): string {
-    return !packageName || packageName[0] !== '@' ? '' : packageName.split(/\/|%2f/)[0];
+  getScope(packageIdent: string): string {
+    // Matches '@scope' in a packageName or pathname, otherwise returns ''
+    if (packageIdent && isScopedPackage(packageIdent)) {
+      const matched = packageIdent.match(/(^@|(?!\/)@).+?((?=%2f)|(?=\/))/);
```
Comment (BYK) [inline src/registries/npm-registry.js]: This regular expression is very hard to follow, can use some comments. Also what's with the `%2f` part? We should unescape the URL and run the matcher on it instead of doing this.

File: __tests__/registries/npm-registry.js [hunk 6]
```diff
@@ -175,6 +175,24 @@ describe('request', () => {
 
     expect(requestParams.headers.authorization).toBe('Bearer testAuthToken');
   });
+
+  test('should add authorization header with correct token for scoped package', () => {
```
Comment (BYK) [inline __tests__/registries/npm-registry.js]: This test case is very similar to the previous one so I'd either make them a bit more differentiated or fix the previous one.

File: src/registries/npm-registry.js (hunk 4 above)
Comment (lukeggchapman) [inline src/registries/npm-registry.js]: You want the complete match (matched[0]), not any of the groups. The first group is '^@' or '/@'

File: src/registries/npm-registry.js (hunk 5 above)
Comment (lukeggchapman) [inline src/registries/npm-registry.js]: I matched on `%2f` because I didn't have the domain knowledge to know for what reason it was escaped in the pathname. If it's not to differentiate them from other forward slashes I could add an extra `.replace('%2f', '/')` call in this line to make the regex more readable.

File: src/registries/npm-registry.js (hunk 3 above)
Comment (lukeggchapman) [inline src/registries/npm-registry.js]: I've reworked the regex so this is possible.

File: __tests__/registries/npm-registry.js (hunk 6 above)
Comment (lukeggchapman) [inline __tests__/registries/npm-registry.js]: I don't believe it is.
This tests that a scope configured with a non-default registry is receiving the token specified for that scope.
The previous test is testing the use of the default _authToken when fetching from a scope that has no configured registry.
I'm happy to remove the previous test if you believe it does not add value.

**Comment by lukeggchapman** [comment]
Thanks for the review @BYK! I've cleaned up some things and It's ready for you to review again.

**Comment by soda0289** [comment]
@lukeggchapman Thanks for fixing this and adding some more tests!

 I have a question about how we determine if a package is scoped and what that scope is.
Using the regex: `/(^@|(?!\/)@).+?((?=%2f)|(?=\/))/`
Won't this mark the following url as having a scope:
`http://my-registry.com/@registry/foo/bar.tgz`
 where registry could be: `http://my-registry.com/@registry/foo`

OR this url having an incorrect scope:
`http://my-registry.com/test@test/@foo/bar.tgz` 
where registry could be `http://my-registry.com/test@test/` and the actual scope `@foo`

Not sure there can be a perfect solution or if it can be improved. Maybe for URL we only selected the last `/` as scope and package name separator but that might not work if package name contains a `/`, which I'm not sure is allowed

**Comment by lukeggchapman** [comment]
@soda0289 You're welcome, happy to contribute.

I've recently refactored the regex to be `/(^@|\/@)(.+?)(?=\/)/` which will resolve the problem with `http://my-registry.com/test@test/@foo/bar.tgz`.

With the first issue I've made the assumption that scoped packages will only ever have one `@/` based off an [existing assumption.](https://github.com/yarnpkg/yarn/blob/master/src/registries/npm-registry.js#L119)

It would be possible to rewrite the regex to grab only the last `@scope/pkg.tgz`, but with registries that have an `@` already always being marked as scoped then more work would need to be done to support them properly.

File: src/registries/npm-registry.js (hunk 4 above)
Comment (BYK) [inline src/registries/npm-registry.js]: Oh, let's make those groups non-capturing then? (haven't seen the latest code yet)

File: src/registries/npm-registry.js (hunk 5 above)
Comment (BYK) [inline src/registries/npm-registry.js]: You need `decodeURIComponent(str)` instead.

**Comment by BYK** [comment]
>It would be possible to rewrite the regex to grab only the last @scope/pkg.tgz, but with registries that have an /@ already always being marked as scoped then more work would need to be done to support them properly

This is a bit concerning. Should we file an issue for this so we don't forget? :(

File: src/registries/npm-registry.js [hunk 7]
```diff
@@ -222,13 +222,9 @@ export default class NpmRegistry extends Registry {
...
   getScope(packageIdent: string): string {
-    // Matches '@scope' in a packageName or pathname, otherwise returns ''
-    if (packageIdent && isScopedPackage(packageIdent)) {
-      const matched = packageIdent.match(/(^@|(?!\/)@).+?((?=%2f)|(?=\/))/);
-      return (matched && matched[0]) || '';
-    }
-
-    return '';
+    // Matches strings beginning with @ or contain /@ up until the next forward slash
+    const match = packageIdent.replace('%2f', '/').match(/(^@|\/@)(.+?)(?=\/)/);
```
Comment (BYK) [inline src/registries/npm-registry.js]: Looking at this closely, I really can't understand a few things:

 - Why do we even have URL-encoded parts in the ident? That doesn't look right so it should be fixed where the value originally comes from.
 - What does the `?` at the end of `(.+?)` part stand for? Is it for the query string delimiter? If so, why do we need to include it in the match?

I feel like the following version is clearer and more accurate:
```
// Matches the first path segment that starts with an `@`. The RegEx is constructed as follows:
// `(?:^|\/)` Match either the start of the string or a `/` but don't capture
// `(@[^?\/]+)` Match a string starting with an `@` and not having any `/` or `?` in it and capture
// `(?:[?\/]|$)/)` Match a path delimiter, a query string delimiter or the end of the string but don't capture
const match = packageIdent.match(/(?:^|\/)(@[^?\/]+)(?:[?\/]|$)/);
return match && match[1] || '';
```

File: __tests__/registries/npm-registry.js [hunk 8]
```diff
@@ -175,6 +175,24 @@ describe('request', () => {
 
     expect(requestParams.headers.authorization).toBe('Bearer testAuthToken');
   });
+
+  test('should add authorization header with token for scope if pathname is to registry and is scoped package', () => {
```
Comment (BYK) [inline __tests__/registries/npm-registry.js]: Let's explicitly say "custom registry" in the test name:
`should add authorization header with token for custom registries with a scoped package`

File: src/registries/npm-registry.js (hunk 7 above)
Comment (lukeggchapman) [inline src/registries/npm-registry.js]: * It's only the forward slashes in the package name (eg (`https://private.npm.repo.com/@foo%2fbar`) and It would be good to find where the pathname is getting stored/retrieved from to see if it is necessary to keep it in that format. I assume it is related to [this](https://github.com/yarnpkg/yarn/blob/master/src/registries/npm-registry.js#L72).
* I'm using the `?` after a quantifier for [non-greedy matching](https://developer.mozilla.org/en/docs/Web/JavaScript/Guide/Regular_Expressions#special-questionmark), so it matches only up to the next forward slash that is in a lookahead so it doesn't get captured.
* I haven't seen a query string in a pathname yet, so it is not something I am catering for. Is that possible? Do you have some examples I could add to the test cases?

Your regex passes the test cases so I'll update it with that, with the added benefit of you having considered the query strings. :+1:

**Comment by lukeggchapman** [comment]
I'll create an issue for registries that contain `/@` always being marked as issues tomorrow.
Also I can look more into where the escaped forward slashes in package names within paths are coming from.

File: __tests__/registries/npm-registry.js [hunk 9]
```diff
@@ -271,7 +271,7 @@ describe('getScope functional test', () => {
     test('in pathname', () => {
       const pathnames = [
         ['http://foo.bar:80/foo/bar/baz', ''],
-        ['http://foo.bar:80/@scopedNoPkg', ''],
+        ['http://foo.bar:80/@scopedNoPkg', '@scopedNoPkg'],
```
Comment (BYK) [inline __tests__/registries/npm-registry.js]: I'm guessing this was intentional?

File: src/registries/npm-registry.js [hunk 10]
```diff
@@ -218,21 +221,29 @@ export default class NpmRegistry extends Registry {
     }
   }
 
-  getScope(packageName: string): string {
-    return !packageName || packageName[0] !== '@' ? '' : packageName.split(/\/|%2f/)[0];
+  getScope(packageIdent: string): string {
+    // removing escaped / from package names in full paths
+    packageIdent = packageIdent.replace('%2f', '/');
```
Comment (BYK) [inline src/registries/npm-registry.js]: I'm trying to say this normalization should not be done here, and actually should not be done at all. `http://foo.bar:80/@scope%2fbar/baz` is not equivalent to `http://foo.bar:80/@scope/bar/baz`. See https://stackoverflow.com/a/14658209/90297

File: __tests__/registries/npm-registry.js (hunk 9 above)
Comment (lukeggchapman) [inline __tests__/registries/npm-registry.js]: Yeah I was on the fence about this one anyway. Semantically getScope is correctly getting the scope, but there is no package and will likely error when trying to fetch the package from the endpoint.

File: src/registries/npm-registry.js (hunk 10 above)
Comment (lukeggchapman) [inline src/registries/npm-registry.js]: Agreed, it is not ideal and smells. Hopefully I'll have some time tomorrow (I'm in Australia) to follow the pathname around and see why it has it in the first place. So far this is the only file I've looked at in the yarn codebase.

**Comment by lukeggchapman** [comment]
So `2%f` was originally introduced when [support for scoped packages was added](https://github.com/lukeggchapman/yarn/commit/1acee606cf21ae725f7aa1ee8802774539f22ec2).

In my case It is being added to the package name when [initially added with `yarn add`](https://github.com/yarnpkg/yarn/blob/master/src/resolvers/registries/npm-resolver.js#L79) and then stored in the path when stored and fetched from the lock file.

Doing a search on the repo for `NpmRegistry.escapeName` will see that it is common practice to replace the forward slash in our package names as this is how the npm registry expects our scoped package names to be formatted.

**Comment by lukeggchapman** [comment]
Knowing that forward slashes in scoped package names are always escaped then I can rewrite both the isScopedPackage check and the getScope method. I'll update the PR shortly.

**Comment by lukeggchapman** [comment]
@BYK Let me know what you think of this latest solution, it's a different approach to the one I was taking before.
@soda0289 latest fix should cover your concerns for registries with `/@`

**Comment by BYK** [comment]
Oh wow, great detective work @lukeggchapman. I'll be honest, I'm not a fan of this latest solution but seems like the "real" solution is not very relevant to this patch and we have passing tests. So I'll merge this and then follow up with a PR in the hopes of fixing this madness.

NPM itself [says the following](https://docs.npmjs.com/misc/scope):
>A scope follows the usual rules for package names (URL-safe characters, no leading dots or underscores). 

So we should already unescape package names when working on them?

File: src/registries/npm-registry.js [hunk 11]
```diff
@@ -23,6 +23,12 @@ const DEFAULT_REGISTRY = 'https://registry.npmjs.org/';
 const REGEX_REGISTRY_PREFIX = /^https?:/;
 const REGEX_REGISTRY_SUFFIX = /registry\/?$/;
 
+// All scoped package names are of the format `@scope%2fpkg` from the use of NpmRegistry.escapeName
+// `(?:^|\/)` Match either the start of the string or a `/` but don't capture
+// `[^\/?]*?` Match any character that is not '/' or '?' and capture, up until the first occurance of:
+// `%2f` Match '%2f' the escaped forward slash and don't capture
+const SCOPED_PKG_REGEXP = /(?:^|\/)(@[^\/?]*?)(?=%2f)/;
```
Comment (BYK) [inline src/registries/npm-registry.js]: This matches an empty scope like `@%2f`. Also, matches `@scope%2f`without the package name. Gonna merge anyways since I don't think this is a problem but I'm quite confused about the `*?` part in `(@[^\/?]*?)`. Why not just `(@[^\/?]+)` instead. What does even `*?` mean since both of them are quantifiers and you're not supposed to use two quantifiers back to back.

_Goes away, reads about `*?` making `*` not greedy._

I'm now convinced that we need `+` instead of `*?` since our aim is to match a scope and the full name of the scope. So gonna replace that and then merge it. LMK if you disagree.

**Comment by lukeggchapman** [comment]
@BYK I agree that the solution is not ideal as it would be much cleaner if the escape name was done only once in one place. If you search for `request(NpmRegistry.escapeName(` it gets used quite frequently.

This sums up my knowledge on the presence of the %2f in npm package names:
https://github.com/npm/npm/issues/11738#issuecomment-190930512
The extensive discussion was a little too extensive for me today 😆

File: src/registries/npm-registry.js (hunk 11 above)
Comment (lukeggchapman) [inline src/registries/npm-registry.js]: I wasn't sure if a eg `@scope%2ffoo%2fbar` should have scope `@scope` or `@scope%2ffoo` so your knowledge is appreciated there.

**Comment by BYK** [comment]
>@BYK I agree that the solution is not ideal as it would be much cleaner if the escape name was done only once in one place. If you search for request(NpmRegistry.escapeName( it gets used quite frequently.

Yup, gonna fix that later on.

>This sums up my knowledge on the presence of the %2f in npm package names:
npm/npm#11738 (comment)
>The extensive discussion was a little too extensive for me today 😆

Yeah, I've also traced back to the weird behavior of NPM registry. I think it is madness but it is the world we live in :D Sorry for all the trouble. Also fixing your PR soon. I messed it up :D

File: src/registries/npm-registry.js (hunk 11 above)
Comment (BYK) [inline src/registries/npm-registry.js]: Turns out we need `+?` so it stops matching at the first sight of `%2f`. TIL

**Comment by lukeggchapman** [comment]
Hey @BYK, sorry to pester you but are you able to give me an indication of when this will be released?
With npm 5 also having [similar issues](https://github.com/npm/npm/issues/17918#issuecomment-318140270) It's a blocker for using private modules in our builds, I'll need to use my fork until it's released.
